### ISSUE TITLE
Problem with git versions

### ISSUE THREAD

**Issue by albestro** [LLNL/spack#36134]
### Steps to reproduce

It seems that #35681 broke something about git versions

```
spack spec -I libtree@git.master
```

### Error message

```console
$ spack spec -I libtree@git.master
==> Error: object of type 'NoneType' has no len()
```

```console
Traceback (most recent call last):
  File "/project/csstaff/ialberto/opt/spack-daint/bin/spack", line 54, in <module>
    sys.exit(main())
  File "/project/csstaff/ialberto/opt/spack-daint/lib/spack/spack_installable/main.py", line 37, in main
    sys.exit(spack.main.main(argv))
  File "/project/csstaff/ialberto/opt/spack-daint/lib/spack/spack/main.py", line 1004, in main
    return _main(argv)
  File "/project/csstaff/ialberto/opt/spack-daint/lib/spack/spack/main.py", line 959, in _main
    return finish_parse_and_run(parser, cmd_name, env_format_error)
  File "/project/csstaff/ialberto/opt/spack-daint/lib/spack/spack/main.py", line 987, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "/project/csstaff/ialberto/opt/spack-daint/lib/spack/spack/main.py", line 636, in _invoke_command
    return_val = command(parser, args)
  File "/project/csstaff/ialberto/opt/spack-daint/lib/spack/spack/cmd/spec.py", line 103, in spec
    concretized_specs = spack.cmd.parse_specs(args.specs, concretize=True)
  File "/project/csstaff/ialberto/opt/spack-daint/lib/spack/spack/cmd/__init__.py", line 222, in parse_specs
    spec.concretize(tests=tests)  # implies normalize
  File "/project/csstaff/ialberto/opt/spack-daint/lib/spack/spack/spec.py", line 2821, in concretize
    self._new_concretize(tests)
  File "/project/csstaff/ialberto/opt/spack-daint/lib/spack/spack/spec.py", line 2796, in _new_concretize
    result = solver.solve([self], tests=tests)
  File "/project/csstaff/ialberto/opt/spack-daint/lib/spack/spack/solver/asp.py", line 2539, in solve
    result, _, _ = self.driver.solve(setup, specs, reuse=reusable_specs, output=output)
  File "/project/csstaff/ialberto/opt/spack-daint/lib/spack/spack/solver/asp.py", line 686, in solve
    setup.setup(self, specs, reuse=reuse)
  File "/project/csstaff/ialberto/opt/spack-daint/lib/spack/spack/solver/asp.py", line 2137, in setup
    self.define_version_constraints()
  File "/project/csstaff/ialberto/opt/spack-daint/lib/spack/spack/solver/asp.py", line 1881, in define_version_constraints
    v for v in sorted(self.possible_versions[pkg_name]) if v.intersects(versions)
  File "/project/csstaff/ialberto/opt/spack-daint/lib/spack/spack/solver/asp.py", line 1881, in <listcomp>
    v for v in sorted(self.possible_versions[pkg_name]) if v.intersects(versions)
  File "/project/csstaff/ialberto/opt/spack-daint/lib/spack/spack/version.py", line 111, in coercing_method
    return getattr(ca, method.__name__)(cb, *args, **kwargs)
  File "/project/csstaff/ialberto/opt/spack-daint/lib/spack/spack/version.py", line 1059, in intersects
    return self.overlaps(other)
  File "/project/csstaff/ialberto/opt/spack-daint/lib/spack/spack/version.py", line 108, in coercing_method
    return method(a, b, *args, **kwargs)
  File "/project/csstaff/ialberto/opt/spack-daint/lib/spack/spack/version.py", line 1050, in overlaps
    if self[s].overlaps(other[o]):
  File "/project/csstaff/ialberto/opt/spack-daint/lib/spack/spack/version.py", line 488, in overlaps
    return self.intersects(other)
  File "/project/csstaff/ialberto/opt/spack-daint/lib/spack/spack/version.py", line 111, in coercing_method
    return getattr(ca, method.__name__)(cb, *args, **kwargs)
  File "/project/csstaff/ialberto/opt/spack-daint/lib/spack/spack/version.py", line 108, in coercing_method
    return method(a, b, *args, **kwargs)
  File "/project/csstaff/ialberto/opt/spack-daint/lib/spack/spack/version.py", line 619, in intersects
    n = min(len(v1), len(v2))
TypeError: object of type 'NoneType' has no len()
```

### Information on your system

Problem seems to start from PR #35681

* **Spack:** 0.20.0.dev0 (99893a64754047f1ad539509fb5b79b780d7477f)
* **Python:** 3.6.15
* **Platform:** cray-sles15-haswell
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by alalazo** [comment]
Reproduced as in the description.

**Comment by alalazo** [comment]
A smaller reproducer:
```console
$ spack python -c 'from spack.spec import Spec; s = Spec("libtree@git.master"); s.intersects("libtree@1.1.0")'
Traceback (most recent call last):
  File "/usr/lib/python3.8/code.py", line 90, in runcode
    exec(code, self.locals)
  File "<input>", line 1, in <module>
  File "/home/culpo/github/spack/lib/spack/spack/spec.py", line 3553, in intersects
    if not self.versions.intersects(other.versions):
  File "/home/culpo/github/spack/lib/spack/spack/version.py", line 1059, in intersects
    return self.overlaps(other)
  File "/home/culpo/github/spack/lib/spack/spack/version.py", line 108, in coercing_method
    return method(a, b, *args, **kwargs)
  File "/home/culpo/github/spack/lib/spack/spack/version.py", line 1050, in overlaps
    if self[s].overlaps(other[o]):
  File "/home/culpo/github/spack/lib/spack/spack/version.py", line 488, in overlaps
    return self.intersects(other)
  File "/home/culpo/github/spack/lib/spack/spack/version.py", line 111, in coercing_method
    return getattr(ca, method.__name__)(cb, *args, **kwargs)
  File "/home/culpo/github/spack/lib/spack/spack/version.py", line 108, in coercing_method
    return method(a, b, *args, **kwargs)
  File "/home/culpo/github/spack/lib/spack/spack/version.py", line 619, in intersects
    n = min(len(v1), len(v2))
TypeError: object of type 'NoneType' has no len()

```

**Comment by alalazo** [comment]
The issue seems to be in the code that downloads the repo and tries to infer a "reference version". A workaround is to supply the reference version explicitly:
```console
$ spack spec -I libtree@git.master=master
```

**Comment by alalazo** [comment]
Lowering the impact since there is a workaround

**Comment by alalazo** [comment]
The issue faced by the user seems to be solved in #35057 , in the sense that the code path taken to infer certain properties for clingo is different. The underlying bug in https://github.com/spack/spack/issues/36134#issuecomment-1471614594 instead stays also with that PR.
